airflow/serialization/serialized_objects.py

```diff
@@ -502,6 +504,8 @@ def deserialize(cls, encoded_var: Any) -> Any:
             return Dataset(**var)
         elif type_ == DAT.SIMPLE_TASK_INSTANCE:
             return SimpleTaskInstance(**cls.deserialize(var))
+        elif type_ == DAT.TASK_INSTANCE:
```

Comment:
Here you are extending legacy code, i suggest using the more generic `serialization` code from `serde`

Reply:
We rely on serialized_objects in InternalAPI in general
https://github.com/apache/airflow/blob/main/airflow/api_internal/internal_api_call.py#L107
Do you think we could switch to serde now? Is it compatible with what serialize_objects offer?

Reply:
Given that you are going to rely on a lot of serialization and the new serializer/deserializer is significantly faster and more future proof (versioning) I think everything that does currently not have a schema (that's everything except a DAG*) should switch.

I am willing to help out to ease the migration if required. 

* I am working on DAG serialization/deserialization but untangling how it is done now and to improve the structure is taking time especially with all the edge cases.

Reply:
Great. I think we will migrate to the new serializer/deserializer in this case, but probably outside of this PR.
If you believe it would better to migrate first, then I can revert this change and get back to it when using new way.
WDYT?

Reply:
Am I incorrect in thinking that migration is a 2 line change in `internal_api_call` and you are not relying on any of the other serialized_objects (basically DAG)? If so then I would say do not add technical dept and migrate now. This allows us to call `serialized_objects` as stale and soon to be deprecated.

Otherwise, keep it and and it to the todo of AIP-44?

Reply:
yes, we need to make change there and server-side:
https://github.com/apache/airflow/blob/main/airflow/api_internal/endpoints/rpc_api_endpoint.py#L76

There are more methods with `internal_api_call` decorator. I did a quick check and I see that (beside primitives) we already need serialization to Dag,DagRun, BaseXCom, CallbackRequest (and probably more soon)

Reply:
Question. (and really sorry I have not looked at it earlier) and also follow up on https://github.com/apache/airflow/pull/29513 as well.

Are we REALLY sure we need to serialize the whole TaskInstance (and other) objects to be passed via internal_api_call ?  

For me this is an indication that we either have too narrow of a scope for an @internal_api call (generally speaking the whole internal_api_call should span the whole DB transaction. And since we are trying to pass an ORM object (TaskInstance, DAGRun etc.) it means that that object must have been retrieved before within a transaction. So it means that our internal_api_call should wrap the retrieval as well. Which might simply mean that we need to do some refactoring and add extract new methods (and then decorate them). 

That's of course a general statement and approach and there might be cases that this require a bit deeper refactoring. 

Which methods are affected @mhenc (besides the https://github.com/apache/airflow/pull/29513 one) ? maybe we can look toghether and figure out approach for all of them ?

Reply:
For example in https://github.com/apache/airflow/pull/29513#discussion_r1106553896 I proposed the solution that would avoid TaskInstance serialization altogether. I am reasonable convinced, that similar approach can be done for all ORM objects of ours and that we do not need to serialize any of them (in which case the whole PR might not be needed).

Reply:
But this if for client -> Internal API side .
What about Internal API -> client, e.g. worker. We need to have TaskInstance object in worker to run the tasks. 
https://github.com/apache/airflow/blob/main/airflow/cli/commands/task_command.py#L187

Reply:
That's **precisely** what I am refactoring now.